app/models/tag.rb

```diff
@@ -1125,49 +1128,79 @@ def remove_favorite_tags
     favorite_tags.destroy_all
   end
 
-  attr_reader :media_string, :fandom_string, :character_string, :relationship_string, :freeform_string, :meta_tag_string, :sub_tag_string, :merger_string
+  attr_reader :meta_tag_string, :sub_tag_string, :merger_string
+
+  # Uses the value of parent_types to determine whether the passed-in tag
+  # should be added as a parent or a child, and then generates the association
+  # (if it doesn't already exist). If it does already exist, returns the
+  # existing CommonTagging object.
+  def build_association(tag)
+    if parent_types.include?(tag&.type)
+      common_taggings.find_or_initialize_by(filterable: tag)
+    else
+      child_taggings.find_or_initialize_by(common_tag: tag)
+    end
+  end
 
-  def add_parent_string(tag_string)
+  # Try to create new associations with the tags of type tag_type whose names
+  # are listed in tag_string.
+  def add_association_string(tag_type, tag_string)
     names = tag_string.split(',').map(&:squish)
     names.each do |name|
       parent = Tag.find_by_name(name)
-      if parent && parent.canonical?
-        add_association(parent)
+      prefix = "Cannot add association to '#{name}':"
+      if parent && parent.type != tag_type
+        errors.add(:base, "#{prefix} #{parent.type} added in #{tag_type} field.")
       else
-        errors.add(:base, "Cannot add association: '#{name}' tag " +
-          (parent ? "is not canonical." : "does not exist."))
+        association = build_association(parent)
+        save_and_gather_errors(association, prefix)
       end
     end
   end
 
-  def fandom_string=(tag_string); self.add_parent_string(tag_string); end
-  def media_string=(tag_string); self.add_parent_string(tag_string); end
-  def character_string=(tag_string); self.add_parent_string(tag_string); end
-  def relationship_string=(tag_string); self.add_parent_string(tag_string); end
-  def freeform_string=(tag_string); self.add_parent_string(tag_string); end
+  # Save an item to the database, if it's valid. If it's invalid, read in the
+  # error messages from the item and copy them over to this tag.
+  def save_and_gather_errors(item, prefix)
+    return unless item.new_record? || item.changed?
+    return if item.valid? && item.save
+
+    item.errors.full_messages.each do |message|
+      errors.add(:base, "#{prefix} #{message}")
+    end
+  end
+
+  # Find and destroy all invalid CommonTaggings and MetaTaggings associated
+  # with this tag.
+  def destroy_invalid_associations
+    common_taggings.destroy_invalid
+    child_taggings.destroy_invalid
+    meta_taggings.destroy_invalid
+    sub_taggings.destroy_invalid
+  end
+
```

Comment:
Just for anyone trying to search for where one of the *_string methods is defined: 
```suggestion
# defines fandom_string=, media_string=, character_string=, relationship_string=, freeform_string= 
```